### PR TITLE
Vue: Support custom blocks

### PR DESCRIPTION
**Pull request by sosukesuzuki**
This improves formatting [custom blocks](https://vue-loader.vuejs.org/guide/custom-blocks.html) in Vue SFC.(see #5448 and #5458)

If there is `lang` attribute, the value of `lang` is passed to `parser` option.(If the value of `lang` is unknown, no formatting is done)

```vue
<!-- Input -->
<custom lang="babel">
const foo =
    'foo'
</custom>

<!-- Output -->
<custom lang="babel">
const foo = "foo";
</custom>
```

**real world custom blocks examples:**

- [vue-i18n(JSON)](https://kazupon.github.io/vue-i18n/guide/sfc.html#basic-usage)
- [vue-i18n(YAML)](https://kazupon.github.io/vue-i18n/guide/sfc.html#yaml-loading)
- [Gridsome(GraphQL)](https://gridsome.org/docs/querying-data/#how-to-query-with-graphql)

How do you think?

- [x] I’ve added tests to confirm my change works.
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/pr-XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Comment by fisker** [comment]
I guess we still can't parse this

```js
<custom lang="babel">
const foo = "<";
</custom>
```

**Comment by sosukesuzuki** [comment]
@fisker Yes, I think we should fix the problem in other way like #7975

**Comment by fisker** [comment]
@sosukesuzuki 
I  think we should add logic here, https://github.com/prettier/prettier/blob/5569e3306ed9b760b44ae48aea8f5626130afa26/src/language-html/utils.js#L364

maybe also both `lang` and `type`? I don't people would like to write `lang="babel"`, should be `lang="json"` or `type="application/json"`.

File: src/language-html/printer-html.js [hunk 1]
```diff
@@ -98,6 +99,30 @@ function embed(path, print, textToDoc, options) {
...
+            printed = textToDoc(node.value, { parser });
+          } catch (error) {
+            // Do nothing
+          }
+        } else {
+          const parsers = ["json", "yaml", "graphql"];
+          for (let i = 0; i < parsers.length && !printed; i++) {
+            const parser = parsers[i];
+            try {
+              printed = textToDoc(node.value, { parser });
```
Comment (fisker) [inline src/language-html/printer-html.js:116]: I don't like the idea guessing languages,  people even complaining about auto format with known language.

Comment (sosukesuzuki) [reply src/language-html/printer-html.js:116]: I don't like this too, but in the real world custom blocks are often used without attributes.

Comment (fisker) [reply src/language-html/printer-html.js:116]: In that way, why don't we ignore it. Want format? Tell me what it is.

Comment (sosukesuzuki) [reply src/language-html/printer-html.js:116]: I think formatting for custom blocks is needed. For example, [eslint-plugin-gridsome](https://github.com/gridsome/eslint-plugin-gridsome/blob/master/lib/rules/format-query-block.js) and [eslint-plugin-prettier-vue](https://github.com/meteorlxy/eslint-plugin-prettier-vue) use Prettier to format custom blocks.
Could you tell me what is the reason for ignoring it?

Comment (fisker) [reply src/language-html/printer-html.js:116]: I'm not saying disable formatting custom blocks.
I'm worried about wrong results.

We should ignore 
```html
<docs></docs>
```

But format
```html
<docs lang="markdown"></docs>
```

Maybe with some exception like `<page-query>` known as `graphql` block, like `<script>` we assume that's `javascript`

Comment (sosukesuzuki) [reply src/language-html/printer-html.js:116]: I understood. What do you think about adding options for custom blocks like [eslint-plugin-prettier-vue](https://github.com/meteorlxy/eslint-plugin-prettier-vue)? (I'm reluctant to add options)

Comment (sosukesuzuki) [reply src/language-html/printer-html.js:116]: I'll remove guessing languages for now. At least for this PR, it only supports custom blocks using the lang / type attribute.

File: src/document/doc-utils.js [hunk 2]
```diff
@@ -182,17 +182,31 @@ function removeLines(doc) {
...
+    const parts = getInnerParts(doc);
+    const lastPart = parts[parts.length - 1];
     if (lastPart.type === "concat") {
       if (
         lastPart.parts.length === 2 &&
         lastPart.parts[0].hard &&
         lastPart.parts[1].type === "break-parent"
       ) {
-        return { type: "concat", parts: doc.parts.slice(0, -1) };
+        return { type: "concat", parts: parts.slice(0, -1) };
```
Comment (evilebottnawi) [inline src/document/doc-utils.js:209]: Potential breaking change for plugins

File: src/language-html/printer-html.js (hunk 1 above)
Comment (fisker) [reply src/language-html/printer-html.js:116]: 👍 , I guess we can use `--embedded-language-formatting` later?

File: src/document/doc-utils.js (hunk 2 above)
Comment (evilebottnawi) [reply src/document/doc-utils.js:209]: I think we can implement the `withInnerParts` argument (and maybe put todo to change it on default for prettier@3)

/cc @thorn0

File: src/language-html/printer-html.js (hunk 1 above)
Comment (sosukesuzuki) [reply src/language-html/printer-html.js]: sounds good.

File: src/document/doc-utils.js (hunk 2 above)
Comment (sosukesuzuki) [reply src/document/doc-utils.js:209]: Thanks. I didn't notice. This change is needed for applying `stripTrailingHardline` to doc printed via `printer-yaml`. I'll try to fix `printer-yaml`.

File: src/language-html/utils.js [hunk 3]
```diff
@@ -416,6 +399,16 @@ function inferScriptParser(node) {
...
 
+  if (options && isCustomBlock(node, options)) {
+    const { lang, type } = node.attrMap;
+    if (type) {
+      if (type.endsWith("json")) {
+        return "json";
+      }
+    }
+    return lang;
+  }
```
Comment (fisker) [inline src/language-html/utils.js:410]: Can we reuse the logic above? so we can support [`type="text/markdown"`](https://deploy-preview-8023--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEAeeBnGACANgQygHMBeAHRAE98BbXCgPjKgSWe23ewAs5dcISDsOEVe-CNgDuEAE64AJgEIKXNVA7MAVvjYaefAUIqBlBkDJDIGsGQIYMgewZAgAyA0OUAyroEB-1VFQB6TDCZRm6HBY2DCUAA5w5CDwAB4wnjT4sgDWChBS-iB+rOoc4kYiImKGkjLyyu6a+uo6enklxiDm1vbObiAB3kG+zAE+2VB1nPr5goWiIKPScooqHdULVbVco43Nto6u7l79vR4+eISkFACuUMlQ6ZkDQysN45qTJdPlc7nDS7p3EmuWG23bLpYPzMEAAGhAEDCMAAltAMMhQElZOkAApJBAIlD4XBSfCUBEQgBGsnwYGScBgAGUwmSYcRkDBZCc4BDuDA6AB1bgwzC0sBwKmY3kwgBuvMoyHAGEJIHpGDgshgqNJRESyAAZjiFRCtBgYgAhUnkylU2hwAAy9Lgmu1rJAepiVPpRFwcAAiicIPBbbgdSBabIFbIpaKWeCA7J6TBOTCFDBuMgABwABghYRRCs5pLCUozQUVoptEIAjl74CqoViQPgMABaFhwBRNiOyOBlmFtlX4NX4X3+hU0GGM5n2jAut2e702pBav32mD4Imx+OJpAAJghTPwMNwLoAwhAaOqUEEAKwRk4KgAqS6xc-9YbgAEkoM3YFSwFHoQBBN9U0I3X7OAAF8QKAA) and `type="text/x-handlebars-template"` etc?

File: src/language-html/printer-html.js [hunk 4]
```diff
@@ -30,6 +30,7 @@ const {
   getPrettierIgnoreAttributeCommentData,
   hasPrettierIgnore,
   inferScriptParser,
+  isCustomBlock,
```
Comment (thorn0) [inline src/language-html/printer-html.js:33]: Please rename it to `isVueCustomBlock`. This parser and printer are also used for HTML and Angular, so Vue-specific things should be marked as such.

**Comment by fisker** [comment]
Just want let you know, I tried use `@vue/component-compiler-utils` to parse `.vue` today, can't get comments at top and can't keep attribute orders, no info in the SFCDescriptor . https://github.com/fisker/prettier/pull/347

File: src/document/doc-utils.js [hunk 5]
```diff
@@ -182,17 +182,31 @@ function removeLines(doc) {
   return mapDoc(doc, removeLinesFn);
 }
 
-function stripTrailingHardline(doc) {
+function getInnerParts(doc) {
+  let { parts } = doc;
+  let lastPart;
+  // Avoid a falsy element like ""
+  for (let i = 0; i < doc.parts.length && !lastPart; i++) {
+    lastPart = parts[parts.length - (i + 1)];
```
Comment (thorn0) [inline src/document/doc-utils.js:190]: I'd rewrite this loop using `i--` instead of `i++`.

**Comment by thorn0** [comment]
**Prettier pr-8023**
[Playground link](https://deploy-preview-8023--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEAeMBXAzjCBbAAgBsBDKAcwF4AdEALwdoD5qpIocCAzCCAygrR4RaAblYEC7TsP6CQwsagD0mHPiYgANCAgAHGAEtoWZKBIAnCxADuABUsJTKEkRskAnqZ0AjCyTAAazgYAGU9AMMKZBgLDDgdAAsYPCIAdUTDeCwIsDhQpyzDADcsj2RwLG8QKKw4Cxg7f3I8EmQuVzqdACssAA8AIX8gkNCSPDgAGSi4ds6EkF6+0KjyIjgARQwIeDmiLpAIizqLCuL47UOLKJg0wwATGETkAA4ABh09azq0-z0Kr5wE7FWY6ACO23gTX0zhAJCwAFooHA4PdUZcLHAIYZMU0SC02kgOvsFnU8IYYnFSat1lsdrMifMdDASD47o9nkgAEzM-yGIirADC+FaFSBAFZLtg4AAVVnOYkHc5wACSUDRsFCYGuBgAgurQjAPOtKfEAL5moA)
```sh
--parser vue
--vue-indent-script-and-style
```

**Input:**
```vue
<custom lang="zzz">
const foo = "foo";
  const foo = "foo";</custom>
```

**Output:**
```vue
<custom lang="zzz"> const foo = "foo";
  const foo = "foo";>

```

**Comment by fisker** [comment]
Why this not caught by `DEEP_COMPARE` test?

**Comment by fisker** [comment]
Sorry, it's not a test fixture...

File: src/document/doc-utils.js (hunk 2 above)
Comment (thorn0) [reply src/document/doc-utils.js]: Are `doc-utils` considered part of the public plugin API?

**Comment by thorn0** [comment]
`lang="babel"` should not be recognized. `lang="javascript"` should. `babel` is not a language, it's Prettier's internal thing.

**Comment by fisker** [comment]
**Prettier pr-8023**
[Playground link](https://deploy-preview-8023--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEAeMBXAzjCBbAAgBsBDKAcwF4AdEAKy2loD5hqoCDaTakCBGdgF9UAekw58zdu3TZchUhRr1GUFmw5cQPZAOGzx8qSAA0ICAAcYAS2hZkoEgCdnEAO4AFFwgcoSRO4kAJ4O5gBGziRgANZwMADKltE2FMgwzhhw5gAWMHhEAOo5NvBYyWBwCb6lNgBupcHI4FhhIKlYcM4wnlHkeCTIAGYBneYMAB4AQlGx8QkkeHAAMqlww6PZqhMJqeREcACKGBDwG0RjIMnOnc7NdVlmV86pMIU2ACYwOcgAHAAM5ksbk6hSilmawLgtzq63MAEcTvBelY-DosABaKBwOAfXFPZxwRE2Qm9Ej9QZIEYXLadPA2dKZWl7A7HU7rKmbcwwEjhd5fH5IABM3KiNiIewAwvgBs1oQBWJ7YOAAFV5fmplwecAAklA8bAEmAXtYAIL6hIwYIHc6dIRCIA)
```sh
--parser vue
```

**Input:**
```vue
<custom lang="json">{
  "a": 1
}</custom>

<custom lang="json">{
  "a": 1
}
</custom>
```

**Output:**
```vue
<custom lang="json"
{
  "a": 1
}
>

<custom lang="json"
{
  "a": 1
}
</custom>

```

**Comment by thorn0** [comment]
Blocks with unknown languages must be preserved as is.

**Prettier pr-8023**
[Playground link](https://deploy-preview-8023--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEAeMBXAzjCBbAAgBsBDKAcwF4AdEALwdoD4BGAJgGZUB6THfJiAA0ICAAcYAS2hZkoEgCcFEAO4AFRQlkoSRFSQCeskQCMFJMAGs4MAMpiLkishgKMcEQAsYeIgHVPSXgsBzA4Wy0gyQA3IINkcCxjECcsOAUYNXNyPBJkADNdNJEAKywADwAhcysbWxI8OAAZJzgCoo8QMvLbJ3IiOABFDAh4dqJikAcFNIUE6PdhKYUnGD9JABMYT2QADgAGETFlNL9zMQTjuFnotpEARxH4LPFtEBIsAFooODgNv6WCjgj0kQKyJByeSQhQmnTSeEkLjccL6A2Goza0I6IhgJBM6y2OyQbBx5kkRD6AGF8LkEtcAKxLbBwAAqeO0MMmCzgAEkoP9YLYwCsJABBfm2GAGAbjNIAXzlQA)
```sh
--parser vue
```

**Input:**
```vue
<custom lang="zzz">123</custom>
```

**Output:**
```vue
<custom lang="zzz">
123
</custom>

```

**Comment by thorn0** [comment]
Please merge `master` or rebase to see how this will work with `--embedded-language-formatting=off`.

**Comment by sosukesuzuki** [comment]
I'm busy this week, so I'll continue working on it over the weekend.

**Comment by sosukesuzuki** [comment]
I worked on this because it wasn't as busy as I thought it would be.

**Comment by thorn0** [comment]
Blocks with unknown languages still aren't preserved.

**Prettier pr-8023**
[Playground link](https://deploy-preview-8023--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEAeMBXAzjCBbAAgBsBDKAcwF4AdEAIwgloD5qoCC2OuPe+CAZozY86JAE6oA9Jhz5WUEABoQEAA4wAltCzJQE8RADuABQkJdKEkSMkAnrpV1xJMAGs4MAMprXmisgw4hhwKgAWMHhEAOphmvBYvmBwXhbxmgBu8XbI4FiOIP5YcOIwJi7keCTIAtbFKgBWWAAeAEIu7p5eJHhwADL+cDV1oSBNzV7+5ERwAIoYEPDDRPUgvuLF4rkZIcpr4v4w0ZoAJjBhyAAcAAwqaobF0S5qufdwmxlDKgCOC-Dl6ksIBIWAAtFA4HATlC9uI4L9NHDyiRKtUkLUVqNinhNIFglipjN5oshuiRioYCQ6MczhckAAmCkuTREKYAYXwVVy7wArHtsHAACpUywY1Y7OAASSg0NgXjABw0AEEZV4YHYZstigBfbVAA)
```sh
--parser vue
```

**Input:**
```vue
<custom lang="boo">
  
  
           foo

  bar</custom>

```

**Output:**
```vue
<custom lang="boo">
foo

  bar
</custom>

```

**Comment by thorn0** [comment]
What about [this point](https://github.com/prettier/prettier/pull/8023#issuecomment-612429446)? Why can that code be parsed in `<script>...</script>` but not in custom blocks? Can we parse custom blocks the same way we parse `script` tags?

**Comment by sosukesuzuki** [comment]
@thorn0 Probably, we cannot fix it from Prettier side. angular-html-parser seems to parse `<script>... </script>` in a different way from `<custom>...</custom>`. However, I think it is possible to ignore the syntax errors that occur in  custom blocks. I'm going to give it a try on other the PR.

**Comment by fisker** [comment]
We can add [this](https://github.com/prettier/prettier/pull/8023#issuecomment-612429446) to `tests/errors` for now.

**Pull request merged**